Add low-precision digamma() and polygamma() functions

fritzo opened the pull request:
This implements `digamma()` and `polygamma()` functions on top of @zou3519's PR #3955 which is currently blocked due to licensing issues. This PR implements a low-precision versions of these functions, but keeps all of Richard's wiring. `polygamma(n,x)` is only implemented for `n=0` (`digamma(x)`) and `n=1` (`trigamma(x)`). This implementation is new code and is unencumbered by a license.

## Why?

Users of torch.distributions are currently blocked from differentiating the density of a number of distributions due to lack of `digamma`: `Chi2`, `Gamma`, `Dirichlet`, and `Beta`. The low-precision `digamma()` in this PR is sufficient accuracy for basic statistical applications.

Accuracy: `digamma()` is accurate to at least 7 digits, `trigamma()` is accurate to at least 5 digits. See [this notebook](https://github.com/fritzo/notebooks/blob/master/polygamma.ipynb) for details.

## Tested

- Updated tests in `test_torch` to test more examples but with lower precision
- Ran all Distributions tests, some of which use `digamma()`
- Ran CUDA tests

zou3519 committed:
Implement digamma

fritzo committed:
Merge remote-tracking branch 'zou3519/digamma' into digamma

fritzo committed:
Implement makeshift finite-difference version if digamma()

fritzo committed:
Remove python polyfill for digamma in torch/distributions/

fritzo committed:
Add more comparison values and fix test_digamma

fritzo commented:
cc @zou3519 @apaszke @gchanan 

Would it be reasonable to merge this and then rebase #3955 while we wait for Cephes support?

fritzo committed:
Fix flake8

fritzo committed:
Fix cuda implementation

zou3519 commented:
Sounds good to me, I can rebase #3955 when the licensing is figured out.

apaszke commented on test/test_torch.py:
```diff
@@ -197,6 +197,48 @@ def test_sinh(self):
+            1.00000000e+01, 3.16227766e+01, 1.00000000e+02,
+            3.16227766e+02, 1.00000000e+03, 3.16227766e+03,
+            1.00000000e+04, 3.16227766e+04,
+        ])
```
Can you compress that to `torch.linspace`?

fritzo replied:
Done. (using `torch.logspace()`)

apaszke commented on test/test_torch.py:
```diff
@@ -197,6 +197,48 @@ def test_sinh(self):
+        if not makeshift:
+            y = torch.Tensor([-10, 0])
+            out = torch.digamma(y)
+            self.assertEqual(out == float('inf'), torch.ByteTensor([1, 1]))
```
nit: `self.assertTrue(out.eq(float('inf')).all())`

apaszke commented on test/test_torch.py:
```diff
@@ -197,6 +197,48 @@ def test_sinh(self):
+        self.assertEqual(out, expected, 1e-5)
+
+        out = x.digamma()
+        self.assertEqual(out, expected, 1e-5)
```
Why not include those three values in the test that compares to scipy below?

fritzo replied:
Leaving these here since I've compressed the values below into a `torch.logspace()`. Also the values below span a wider range and require a lower precision (since `TestCase.assertEqual()` does not deal well with relative error).

fritzo commented on test/test_cuda.py:
```diff
@@ -1093,6 +1094,23 @@ def test_var_stability(self):
+            self.assertEqual(norm_errors, zeros)
+
+        test(True)
+        test(False)
```
Quoting Richard from the PR on which this is based #3955 :
> The values of digamma on gpu when compared to cpu are the same to 6 digits. I tried making this a normal test and some output values were high (~100), resulting in an error of around 1e-3, which is above the error threshold

It's probably out of scope for this PR, but maybe you could change the generic tests to assert on relative error rather than absolute error?

apaszke replied:
We have [means to control the required precision](https://github.com/probtorch/pytorch/blob/978b197bc8bbe44e1946367a763fd54c1ed55540/test/test_cuda.py#L326-L335), so it would be better to use that.

fritzo replied:
meh it's easier to just implement a more accurate `digamma()` :wink:

fritzo replied:
Ok I've set custom precisions as you suggest, however they're so lax (1e0 and 1e3) that I've kept @zou3519's the relative-error tests.

fritzo committed:
Updates per review

fritzo committed:
Implement makeshift trigamma

fritzo committed:
Implement higher-precision versions of digamma,trigamma

fritzo committed:
Merge branch 'trigamma' into makeshift-digamma

fritzo committed:
Implement higher-precision CUDA digamma,trigamma

fritzo committed:
Merge branch 'trigamma' into makeshift-digamma

fritzo committed:
Fix compile errors in CUDA digamma, trigamma

fritzo commented on aten/src/THC/THCTensorMathPointwise.cuh:
```diff
@@ -779,5 +779,49 @@ struct TensorBitXorOp {
 };
 
+template <typename real, typename accreal>
+struct TensorDigammaOp {
```
@zou3519 could you help me get this CUDA version working? I tried to translate the working C version, but `test/test_cuda.py::TestCuda::test_FloatTensor_digamma` is failing and I think I may have missed some subtleties of CUDA.
<details>

```
_________________________________________ TestCuda.test_FloatTensor_digamma __________________________________________

self = <test_cuda.TestCuda testMethod=test_FloatTensor_digamma>

    def tmp(self):
        cpu_tensor = tensor_constructor(t)
        type_map = {}
        if force_gpu_half:
            type_map = {
                'torch.FloatTensor': 'torch.cuda.HalfTensor',
                'torch.DoubleTensor': 'torch.cuda.HalfTensor',
            }
        gpu_tensor = to_gpu(cpu_tensor, type_map)
        cpu_args = arg_constructor(t)
        gpu_args = [to_gpu(arg, type_map) for arg in cpu_args]
        cpu_result = getattr(cpu_tensor, fn)(*cpu_args)
        try:
            gpu_result = getattr(gpu_tensor, fn)(*gpu_args)
        except RuntimeError as e:
            reason = e.args[0]
            if 'only supports floating-point types' in reason or 'unimplemented data type' in reason:
                raise unittest.SkipTest('unimplemented data type')
            raise
        except AttributeError as e:
            reason = e.args[0]
            if 'object has no attribute' in reason:
                raise unittest.SkipTest('unimplemented data type')
            raise
        # If one changes, another should change as well
        self.assertEqual(cpu_tensor, gpu_tensor, precision)
        self.assertEqual(cpu_args, gpu_args, precision)
        # Compare results
>       self.assertEqual(cpu_result, gpu_result, precision)

test/test_cuda.py:421:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/common.py:228: in assertEqual
    assertTensorsEqual(x, y)
test/common.py:220: in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
E   AssertionError: 0.185546875 not less than or equal to 1e-05
```
</details>

fritzo replied:
Never mind, this was just the same issue you had with large values causing large absolute error but low relative error.

fritzo commented:
@apaszke I've cleaned up the `test_torch.py` tests so that `digamma` and `trigamma` now have simple tests like other math functions. I also switched to a simple Stirling approximation rather than finite difference. We should still swap in @zou3519's Cephes port for `digamma()` when we get permission since this PR doesn't correctly handle INFINITY at poles and is less precise.

fritzo committed:
Set custom precision for digamma,trigamma in cuda tests

fritzo committed:
Add test_trigamma that tests for relative error

apaszke commented:
@pytorchbot add to whitelist

apaszke reviewed (changes requested):
@fritzo you've accidentally committed ninja build files.

fritzo committed:
Remove accidentally committed ninga build files

fritzo commented:
Thanks, I've removed the build files.

fritzo committed:
Fix type conversion in cuda digamma

apaszke commented on aten/src/TH/THMath.h:
```diff
@@ -126,4 +126,70 @@ Date:  February 1996
+    x += 1;
+  }
+  const float ixx = 1 / (x*x);
+  result += (1 + 1 / (2*x) + ixx * (1./6 - ixx * (1./30 - ixx * (1./42)))) / x;
```
nit: `1./6` will force conversion to double, and I don't know if that's desired. I think it's fine to have only the double overloads unless this function gets auto-vectorized by the compiler (you can check that in objdump). Scalar double math is the same speed as float math on CPUs, and looking at the body of this function I doubt that the casts will become a bottleneck. For vector instructions you get 2x throughput, so it matters.

fritzo replied:
Sorry my cuda skills are weak. How can I specify floating point constant of type `real`? I wrote one version that cast constants everywhere like `(real)(1./6)` but that seemed overly complex.

fritzo replied:
oh, I see :blush: this is C, not CUDA. Done.

apaszke replied:
An easier way is to do `1.f` - this declares a float literal (`1.` is a double literal)

apaszke commented on aten/src/THC/THCTensorMathPointwise.cuh:
```diff
@@ -779,5 +779,49 @@ struct TensorBitXorOp {
+  operator()(real* out, real* in) {
+    static accreal PI = 3.14159265358979323846;
+    accreal x = ScalarConvert<real, accreal>::to(*in);
+    accreal result = 0;
```
You probably don't want to do all this math in `accreal`. This will have a huge perf penalty on non-Tesla cards

fritzo replied:
Is there a way to use `real` for everything but `halffloat` and use `float` for `halffloat`? I originally implemented the all-`real` solution you suggest but there were a bunch of `halffloat` errors I didn't understand and template specialization seemed overly complex.

apaszke replied:
One way would be to do some template magic a'la
```cpp
using compute_type = typename std::conditional<std::is_same<real, half>::value, accreal, real>::type;
```

fritzo replied:
Done.

apaszke commented on test/test_torch.py:
```diff
@@ -197,6 +197,16 @@ def test_sinh(self):
+
+    @unittest.skipIf(not TEST_SCIPY, "Scipy not found")
+    def test_trigamma(self):
+        from scipy.special import polygamma
```
I've also started to wonder if we really want to expose trigamma instead of a more general `polygamma`. Can you change it to be consistent with SciPy?

fritzo replied:
I like the idea of implementing `polygamma()` and I've looked into it, however it's much more complicated than `trigamma()`. Would you prefer a `polygamma(n,x)` that simply errors at `n>=2`?

fritzo replied:
Note that Pymc3 has a `trigamma()` ([code](https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/special.py#L7)). If you do prefer `polygamma()` could you give me some tips on out how to change this PR? It was easy for me to implement `trigamma(x)` since I could copy wiring from @zou3519's `digamma(x)`, but `polygamma(n,x)` is a more complex interface and I'm not sure whether it will fit into the generic C macros and python tests.

apaszke replied:
Yes I've been thinking of having a polygamma that only supports `n` up to 2 (or up to `1`, if we want to be consistent with SciPy). It should still be quite straightforward - just add an extra argument in all yaml declarations (`- int degree`), and add it to the C signature

fritzo replied:
Ok thanks, I'll give that a try.

apaszke replied:
Oh, degree might be a bad idea if we go with SciPy convention

fritzo replied:
Ok I've updated this PR to expose `polygamma(n,x)` instead of `trigamma(x)`. CPU stuff works; I'm still working on CUDA stuff.

apaszke reviewed (commented):
This looks good, but I'd rather limit the interface to `lgamma`, `digamma` and `polygamma`, so it won't keep growing. Also two minor comments that you might want to fix before merging

fritzo committed:
Avoid double-precision arithmetic in single-precision TH_trigammaf()

fritzo committed:
Use single precision in CUDA digamma and trigamma

fritzo committed:
Merge branch 'makeshift-digamma' of github.com:probtorch/pytorch into makeshift-digamma

apaszke committed:
Add const

fritzo committed:
Expose polygamma(n,x); hide trigamma(x)

fritzo committed:
Fix CUDA implementation of polygamma(n,x)

fritzo commented:
Ok `torch.polygamma(n,x)` now works for `n in [0,1]` :smile: Thanks for the help @apaszke!

apaszke commented:
Phew, thanks @fritzo!
